File: ios/Cellule/Ecrans/EcranEstimer.swift
import SwiftUI

struct EcranEstimer: View {

    @EnvironmentObject private var reglages: Reglages
    @EnvironmentObject private var etat: EtatApplication

    @State private var ciel: EtatDuCiel = CIELS[0]
    @State private var obstacle: Obstacle = OBSTACLES[0]
    @State private var renvois: Set<String> = []
    @State private var modele: Modele = .physique
    @State private var parCalcul = true

    @State private var lieu: Lieu = LIEUX[0]
    @State private var mois: Int = Calendar.current.component(.month, from: Date())
    @State private var heure: Double = {
        let m = Calendar.current.dateComponents([.hour, .minute], from: Date())
        return Double(m.hour ?? 12) + Double(m.minute ?? 0) / 60.0
    }()
    @State private var hauteurAVue: Double = 45.0

    private var jour: Int {
        Calendrier.jourDeLAnnee(
            annee: Calendar.current.component(.year, from: Date()),
            mois: mois,
            jour: 15
        )
    }

    private var hauteur: Double {
        guard parCalcul else { return hauteurAVue }
        return Soleil.hauteur(
            latitudeNord: lieu.latitude,
            jourDeLAnnee: jour,
            tempsSolaire: Soleil.tempsSolaire(
                heureLegale: heure,
                jourDeLAnnee: jour,
                longitudeEst: lieu.longitude,
                decalageFuseau: reglages.fuseau
            )
        )
    }

    private var estimation: Estimation {
        Estimateur.calculer(
            Scene(
                ciel: ciel,
                hauteurSoleil: hauteur,
                obstacle: obstacle,
                renvois: renvois,
                iso: reglages.iso,
                modele: modele
            )
        )
    }

    /// « 17h30 » — l'heure du curseur, telle qu'on la lit.
    private var heureLisible: String {
        let heures = Int(heure)
        let minutes = Int(heure.truncatingRemainder(dividingBy: 1) * 60)
        return String(format: "%02dh%02d", heures, minutes)
    }

    /// Où en est le soleil : son midi, son culmen, et sa hauteur à cette heure.
    private var phraseSoleil: String {
        let midi = Soleil.midiSolaire(
            jourDeLAnnee: jour, longitudeEst: lieu.longitude, decalageFuseau: reglages.fuseau
        )
        let midiLisible = String(
            format: "%dh%02d",
            Int(midi),
            arrondi(midi.truncatingRemainder(dividingBy: 1) * 60)
        )
        let culmen = fmt(
            Soleil.hauteur(latitudeNord: lieu.latitude, jourDeLAnnee: jour, tempsSolaire: 12.0)
        )
        return "Midi solaire à \(midiLisible) · le soleil culmine à \(culmen)°"
            + " · à l'heure indiquée il est à \(fmt(hauteur))°."
    }

    private var vitesseVoulue: Double {
        reglages.vitesseReference > 0 ? reglages.vitesseReference : 1.0 / Double(reglages.iso)
    }

    var body: some View {
        let e = estimation
        let couple = coupleConseille(evAppareil: e.evAppareil, vitesseVoulueSec: vitesseVoulue)

        ScrollView {
            VStack(spacing: 0) {
                Spacer().frame(height: Espace.interligne)
                resultat(e: e, couple: couple)
                carteCiel(e: e)
                carteHauteur
                carteObstacle(e: e)
                carteRenvois(e: e)
                carteFilm(e: e)
            }
            .padding(.bottom, 16)
        }
        .background(Teinte.fond)
        .onAppear {
            lieu = LIEUX.min { abs($0.latitude - reglages.latitude) < abs($1.latitude - reglages.latitude) } ?? LIEUX[0]
            partager(e: estimation)
        }
        .onChange(of: e.ev100Scene) { _ in partager(e: estimation) }
    }

    /// Ce que le carnet reprendra si on lui verse l'estimation.
    private func partager(e: Estimation) {
        etat.derniereEstimation = e.ev100Scene
        etat.dernierTypeScene = Statistiques.typeSuggere(ev100: e.ev100Scene, obstacle: obstacle)
        let chaine = e.termes
            .filter { $0.diaphs != nil }
            .map { terme -> String in
                let valeur: String = signe(terme.diaphs ?? 0)
                let nom: String = terme.nom.lowercased()
                return "\(valeur) \(nom)"
            }
            .joined(separator: ", ")
        let debut: String = "ciel \(fmt(e.ev100Ciel))"
        let milieu: String = chaine.isEmpty ? "" : ", \(chaine)"
        etat.derniereChaine = "\(debut)\(milieu) → \(fmt(e.ev100Scene))"
    }

    private func resultat(e: Estimation, couple: Couple?) -> some View {
        Carte {
            HStack(alignment: .lastTextBaseline, spacing: 0) {
                Text(fmt(e.ev100Scene))
                    .font(Police.chiffreEnorme)
                    .foregroundColor(Teinte.surSurface)
                Text("  EV₁₀₀")
                    .font(Police.chiffreMoyen)
                    .foregroundColor(Teinte.surVarianteSurface)
            }
            if let couple = couple {
                Text("\(libelleOuverture(couple.ouverture)) · \(couple.vitesse.libelle)   à \(reglages.iso) ISO")
                    .font(.system(size: 24, weight: .semibold).monospacedDigit())
                    .foregroundColor(Teinte.primaire)
                Text("Exact \(libelleOuverture(couple.ouvertureExacte)) — l'arrondi coûte \(libelleDiaphs(couple.ecartDiaphs)).")
                    .font(Police.detail)
                    .foregroundColor(Teinte.surVarianteSurface)
            }
            Spacer().frame(height: 10)
            Rectangle().fill(Teinte.contourVariante).frame(height: 1)
            Spacer().frame(height: 6)
            ForEach(e.termes) { terme in
                Ligne(
                    intitule: terme.nom,
                    valeur: terme.ev.map { "EV " + fmt($0) } ?? signe(terme.diaphs ?? 0),
                    detail: terme.detail,
                    accent: terme.ev != nil
                )
            }
            Text(
                modele == .physique
                    ? "Les termes s'additionnent en diaphs, jamais en pourcentages — c'est tout l'intérêt du log₂."
                    : "Grille mentale : facteurs bruts, exactement ce que tu calcules de tête."
            )
            .font(Police.detail)
            .foregroundColor(Teinte.surVarianteSurface)
            .fixedSize(horizontal: false, vertical: true)
            .padding(.top, 8)
        }
    }

    private func carteCiel(e: Estimation) -> some View {
        Carte(
            titre: "1 · La lumière du ciel",
            sousTitre: "Le repère infaillible, c'est l'ombre portée."
        ) {
            ForEach(CIELS) { c in
                let ev = modele == .physique
                    ? Ciel.eclairement(hauteurDeg: hauteur, ciel: c).ev100
                    : 15 + c.plat + Ciel.perteHauteurPlate(hauteur)
                Option(
                    nom: c.nom, detail: c.detail,
                    valeur: "EV " + fmt(ev),
                    actif: c.id == ciel.id
                ) { ciel = c }
            }
        }
    }

    private var carteHauteur: some View {
        Carte(
            titre: "2 · La hauteur du soleil",
            sousTitre: "L'éclairement suit le sinus de la hauteur, puis chute plus vite encore sous 15°, l'atmosphère traversée s'épaississant."
        ) {
            ChoixSegmente(
                options: [true, false],
                selection: $parCalcul,
                libelle: { $0 ? "Calculer" : "À vue" }
            )
            Spacer().frame(height: 6)
            if parCalcul {
                LigneBoutons {
                    Deroulant(
                        options: LIEUX,
                        selection: $lieu,
                        libelle: { $0.nom },
                        surChoix: { l in
                            reglages.latitude = l.latitude
                            reglages.longitude = l.longitude
                        }
                    )
                    Deroulant(
                        options: Array(1...12),
                        selection: $mois,
                        libelle: { moisNom($0) }
                    )
                }
                Spacer().frame(height: 6)
                Text("Heure légale : \(heureLisible)")
                    .font(Police.corps)
                    .foregroundColor(Teinte.surSurface)
                Slider(value: $heure, in: 0...23.98)
                    .tint(Teinte.primaire)
                Text(phraseSoleil)
                .font(Police.detail)
                .foregroundColor(Teinte.surVarianteSurface)
                .fixedSize(horizontal: false, vertical: true)
            } else {
                ForEach(RepereOmbre.tous) { repere in
                    let ev = Ciel.eclairement(hauteurDeg: repere.hauteur, ciel: ciel).ev100
                    Option(
                        nom: "\(Int(repere.hauteur))°", detail: repere.detail,
                        valeur: signe(ev - 15),
                        actif: hauteurAVue == repere.hauteur
                    ) { hauteurAVue = repere.hauteur }
                }
            }
        }
    }

    private func carteObstacle(e: Estimation) -> some View {
        Carte(
            titre: "3 · Ce qui bouche la lumière",
            sousTitre: "L'obstacle dominant entre le ciel et ton sujet."
        ) {
            ForEach(OBSTACLES) { o in
                let v = modele == .physique ? o.diaphs(partSoleil: e.partSoleil) : o.auSoleil
                Option(
                    nom: o.nom, detail: o.detail,
                    valeur: abs(v) < 0.05 ? "—" : signe(v),
                    actif: o.id == obstacle.id
                ) { obstacle = o }
            }
        }
    }

    private func carteRenvois(e: Estimation) -> some View {
        Carte(
            titre: "4 · Ce qui la renvoie",
            sousTitre: "Cumulable. Des gains d'éclairement réel, pas une correction de mesure."
        ) {
            let facteur = modele == .physique ? facteurRenvoi(partSoleil: e.partSoleil) : 1.0
            ForEach(RENVOIS) { r in
                Option(
                    nom: r.nom,
                    valeur: signe(r.diaphs * facteur),
                    actif: renvois.contains(r.id)
                ) {
                    if renvois.contains(r.id) { renvois.remove(r.id) } else { renvois.insert(r.id) }
                }
            }
        }
    }

    private func carteFilm(e: Estimation) -> some View {
        Carte(titre: "5 · Le film dans le dos") {
            LigneBoutons {
                Deroulant(
                    options: SENSIBILITES,
                    selection: $reglages.iso,
                    libelle: { "\($0) ISO" }
                )
                Deroulant(
                    options: [Modele.physique, Modele.grilleMentale],
                    selection: $modele,
                    libelle: { $0 == .physique ? "Physique" : "Grille mentale" }
                )
            }
            Text(
                modele == .physique
                    ? "Le modèle physique tient compte de la masse d'air et du ciel diffus : par temps couvert la hauteur du soleil compte beaucoup moins, et une ombre ne coupe que du soleil qui existe."
                    : "La grille mentale reprend les facteurs plats du document de référence. L'écart avec le modèle physique te dit où l'approximation lâche."
            )
            .font(Police.detail)
            .foregroundColor(Teinte.surVarianteSurface)
            .fixedSize(horizontal: false, vertical: true)
            .padding(.top, 8)
            Spacer().frame(height: 8)
            BoutonPlat(texte: "Verser cette estimation au carnet", accent: true, extensible: true) {
                etat.annonceProposee = e.ev100Scene
            }
        }
    }
}


/// L'ombre portée comme goniomètre : ce qu'on lit au sol, et la hauteur qu'il vaut.
struct RepereOmbre: Identifiable {
    let hauteur: Double
    let detail: String
    var id: Double { hauteur }

    static let tous: [RepereOmbre] = [
        RepereOmbre(hauteur: 75.0, detail: "Ombre très courte, été, midi"),
        RepereOmbre(hauteur: 60.0, detail: "Ombre = ⅗ de ta taille"),
        RepereOmbre(hauteur: 45.0, detail: "Ombre = ta taille exactement"),
        RepereOmbre(hauteur: 30.0, detail: "Ombre = 2 × ta taille"),
        RepereOmbre(hauteur: 20.0, detail: "Ombre = 3 × ta taille"),
        RepereOmbre(hauteur: 10.0, detail: "Ombre = 6 × ta taille, lumière rasante"),
        RepereOmbre(hauteur: 4.0, detail: "Soleil sur l'horizon"),
        RepereOmbre(hauteur: -3.0, detail: "Crépuscule civil")
    ]
}

File: app/src/main/kotlin/fr/cellule/app/ecrans/EcranEstimer.kt
package fr.cellule.app.ecrans

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import fr.cellule.app.EtatApplication
import fr.cellule.app.Reglages
import fr.cellule.app.ZoneNavFlottante
import fr.cellule.core.CIELS
import fr.cellule.core.Estimateur
import fr.cellule.core.LIEUX
import fr.cellule.core.Modele
import fr.cellule.core.OBSTACLES
import fr.cellule.core.Photometrie
import fr.cellule.core.RENVOIS
import fr.cellule.core.SENSIBILITES
import fr.cellule.core.Scene
import fr.cellule.core.Soleil
import fr.cellule.core.Statistiques
import fr.cellule.core.coupleConseille
import fr.cellule.core.libelleDiaphs
import fr.cellule.core.libelleOuverture
import java.time.LocalDate
import java.time.LocalTime

/** Une ligne d'option sélectionnable, avec sa valeur en diaphs à droite. */
@Composable
private fun Option(nom: String, detail: String, valeur: String, actif: Boolean, surClic: () -> Unit) {
    Row(
        Modifier
            .fillMaxWidth()
            .padding(vertical = 2.dp)
            .background(
                if (actif) MaterialTheme.colorScheme.primaryContainer
                else MaterialTheme.colorScheme.surfaceVariant,
                RoundedCornerShape(9.dp)
            )
            .clickable { surClic() }
            .padding(horizontal = 11.dp, vertical = 9.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Column(Modifier.weight(1f)) {
            Text(
                nom,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = if (actif) FontWeight.Medium else FontWeight.Normal
            )
            if (detail.isNotBlank()) {
                Text(
                    detail,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
        Text(
            valeur,
            style = MaterialTheme.typography.bodySmall,
            color = if (actif) MaterialTheme.colorScheme.primary
            else MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

@Composable
fun EcranEstimer(reglages: Reglages, etat: EtatApplication) {

    var ciel by remember { mutableStateOf(CIELS[0]) }
    var obstacle by remember { mutableStateOf(OBSTACLES[0]) }
    var renvois by remember { mutableStateOf(setOf<String>()) }
    var modele by remember { mutableStateOf(Modele.PHYSIQUE) }
    var parCalcul by remember { mutableStateOf(true) }

    var lieu by remember {
        mutableStateOf(LIEUX.minByOrNull { kotlin.math.abs(it.latitude - reglages.latitude) } ?: LIEUX[0])
    }
    var mois by remember { mutableStateOf(LocalDate.now().monthValue) }
    val maintenant = remember { LocalTime.now() }
    var heure by remember { mutableStateOf(maintenant.hour + maintenant.minute / 60f) }
    var hauteurAVue by remember { mutableStateOf(45.0) }

    val jour = LocalDate.of(LocalDate.now().year, mois, 15).dayOfYear
    val hauteur = if (parCalcul) {
        Soleil.hauteur(
            lieu.latitude, jour,
            Soleil.tempsSolaire(heure.toDouble(), jour, lieu.longitude, reglages.fuseau)
        )
    } else hauteurAVue

    val scene = Scene(ciel, hauteur, obstacle, renvois, reglages.iso, modele)
    val estimation = Estimateur.calculer(scene)
    etat.derniereEstimation = estimation.ev100Scene
    etat.dernierTypeScene = Statistiques.typeSuggere(estimation.ev100Scene, obstacle)
    etat.derniereChaine = estimation.termes
        .filter { it.diaphs != null }
        .joinToString(", ") { signe(it.diaphs!!) + " " + it.nom.lowercase() }
        .let { "ciel " + fmt(estimation.ev100Ciel) + (if (it.isBlank()) "" else ", $it") +
            " → " + fmt(estimation.ev100Scene) }

    val vitesseVoulue =
        if (reglages.vitesseReference > 0) reglages.vitesseReference else 1.0 / reglages.iso
    val couple = coupleConseille(estimation.evAppareil, vitesseVoulue)

    Column(
        Modifier.fillMaxWidth().verticalScroll(rememberScrollState()).padding(bottom = ZoneNavFlottante)
    ) {

        Carte {
            Row(verticalAlignment = Alignment.Bottom) {
                Text(
                    fmt(estimation.ev100Scene),
                    fontSize = 52.sp,
                    fontWeight = FontWeight.SemiBold
                )
                Text(
                    "  EV₁₀₀",
                    style = MaterialTheme.typography.titleMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.padding(bottom = 8.dp)
                )
            }
            if (couple != null) {
                Text(
                    "${libelleOuverture(couple.ouverture)} · ${couple.vitesse.libelle}   à ${reglages.iso} ISO",
                    fontSize = 24.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = MaterialTheme.colorScheme.primary
                )
                Text(
                    "Exact ${libelleOuverture(couple.ouvertureExacte)} — l'arrondi coûte ${libelleDiaphs(couple.ecartDiaphs)}.",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
            Spacer(Modifier.height(10.dp))
            HorizontalDivider()
            Spacer(Modifier.height(6.dp))
            estimation.termes.forEach { terme ->
                Ligne(
                    intitule = terme.nom,
                    valeur = terme.ev?.let { "EV " + fmt(it) } ?: signe(terme.diaphs ?: 0.0),
                    detail = terme.detail,
                    accent = terme.ev != null
                )
            }
            Text(
                if (modele == Modele.PHYSIQUE)
                    "Les termes s'additionnent en diaphs, jamais en pourcentages — c'est tout l'intérêt du log₂."
                else
                    "Grille mentale : facteurs bruts, exactement ce que tu calcules de tête.",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(top = 8.dp)
            )
        }

        Carte(titre = "1 · La lumière du ciel", sousTitre = "Le repère infaillible, c'est l'ombre portée.") {
            CIELS.forEach { c ->
                val ev = if (modele == Modele.PHYSIQUE)
                    fr.cellule.core.Ciel.eclairement(hauteur, c).ev100
                else 15 + c.plat + fr.cellule.core.Ciel.perteHauteurPlate(hauteur)
                Option(c.nom, c.detail, "EV " + fmt(ev), c.id == ciel.id) { ciel = c }
            }
        }

        Carte(
            titre = "2 · La hauteur du soleil",
            sousTitre = "L'éclairement suit le sinus de la hauteur, puis chute plus vite encore sous 15°, l'atmosphère traversée s'épaississant."
        ) {
            ChoixSegmente(
                options = listOf(true, false),
                selection = parCalcul,
                libelle = { if (it) "Calculer" else "À vue" },
                surChoix = { parCalcul = it }
            )
            Spacer(Modifier.height(6.dp))
            if (parCalcul) {
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    Deroulant("", LIEUX, lieu, { it.nom }) {
                        lieu = it
                        reglages.latitude = it.latitude
                        reglages.longitude = it.longitude
                    }
                    Deroulant("", (1..12).toList(), mois, { moisNom(it) }) { mois = it }
                }
                Spacer(Modifier.height(6.dp))
                Text(
                    "Heure légale : " + String.format("%02dh%02d", heure.toInt(), ((heure % 1) * 60).toInt()),
                    style = MaterialTheme.typography.bodyMedium
                )
                Slider(value = heure, onValueChange = { heure = it }, valueRange = 0f..23.98f)
                val midi = Soleil.midiSolaire(jour, lieu.longitude, reglages.fuseau)
                Text(
                    "Midi solaire à " + String.format("%dh%02d", midi.toInt(), Math.round((midi % 1) * 60)) +
                        " · le soleil culmine à " + fmt(Soleil.hauteur(lieu.latitude, jour, 12.0)) +
                        "° · à l'heure indiquée il est à " + fmt(hauteur) + "°.",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            } else {
                listOf(
                    75.0 to "Ombre très courte, été, midi",
                    60.0 to "Ombre = ⅗ de ta taille",
                    45.0 to "Ombre = ta taille exactement",
                    30.0 to "Ombre = 2 × ta taille",
                    20.0 to "Ombre = 3 × ta taille",
                    10.0 to "Ombre = 6 × ta taille, lumière rasante",
                    4.0 to "Soleil sur l'horizon",
                    -3.0 to "Crépuscule civil"
                ).forEach { (h, detail) ->
                    val ev = fr.cellule.core.Ciel.eclairement(h, ciel).ev100
                    Option("${h.toInt()}°", detail, signe(ev - 15), hauteurAVue == h) { hauteurAVue = h }
                }
            }
        }

        Carte(titre = "3 · Ce qui bouche la lumière", sousTitre = "L'obstacle dominant entre le ciel et ton sujet.") {
            OBSTACLES.forEach { o ->
                val v = if (modele == Modele.PHYSIQUE) o.diaphs(estimation.partSoleil) else o.auSoleil
                Option(
                    o.nom, o.detail,
                    if (kotlin.math.abs(v) < 0.05) "—" else signe(v),
                    o.id == obstacle.id
                ) { obstacle = o }
            }
        }

        Carte(
            titre = "4 · Ce qui la renvoie",
            sousTitre = "Cumulable. Des gains d'éclairement réel, pas une correction de mesure."
        ) {
            val facteur = if (modele == Modele.PHYSIQUE)
                fr.cellule.core.facteurRenvoi(estimation.partSoleil) else 1.0
            RENVOIS.forEach { r ->
                Option(r.nom, "", signe(r.diaphs * facteur), r.id in renvois) {
                    renvois = if (r.id in renvois) renvois - r.id else renvois + r.id
                }
            }
        }

        Carte(titre = "5 · Le film dans le dos") {
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                Deroulant("", SENSIBILITES, reglages.iso, { "$it ISO" }) { reglages.iso = it }
                Deroulant(
                    "", listOf(Modele.PHYSIQUE, Modele.GRILLE_MENTALE), modele,
                    { if (it == Modele.PHYSIQUE) "Physique" else "Grille mentale" }
                ) { modele = it }
            }
            Text(
                if (modele == Modele.PHYSIQUE)
                    "Le modèle physique tient compte de la masse d'air et du ciel diffus : par temps couvert la hauteur du soleil compte beaucoup moins, et une ombre ne coupe que du soleil qui existe."
                else
                    "La grille mentale reprend les facteurs plats du document de référence. L'écart avec le modèle physique te dit où l'approximation lâche.",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(top = 8.dp)
            )
            Spacer(Modifier.height(8.dp))
            BoutonPlat("Verser cette estimation au carnet", accent = true) {
                etat.annonceProposee = estimation.ev100Scene
            }
        }
    }
}

fun moisNom(m: Int): String = listOf(
    "janvier", "février", "mars", "avril", "mai", "juin",
    "juillet", "août", "septembre", "octobre", "novembre", "décembre"
)[(m - 1).coerceIn(0, 11)]
